EIPS/eip-2003.md

```diff
@@ -0,0 +1,116 @@
+
+1. Advertise the [`EVMC_CAPABILITY_PRECOMPILES`] capability
+   in the [`get_capabilities()`] method.
+2. Implement the [`execute()`] method in the following way:
+   1. Check if the call destination address ([`evmc_message::destination`])
+      is within the range of precompiled contracts defined by [EIP-1352].
+      If not, abort execution with the [`EVMC_REJECTED`] status code.
+   2. Check if the call destination address ([`evmc_message::destination`])
+      targets existing precompiled contract.
+      Consider the EVM revision ([`evmc_revision`]) requested by
+      the `rev` parameter of [`execute()`].
+      If yes, execute as follows:
+      1. Inspect the input data ([`evmc_message::input_data`], [`evmc_message::input_size`])
+         and calculate the _gas cost_ of the execution.
+      2. Compute the amount of _gas left_ after execution by
+         subtracting the _gas cost_ from the call gas limit ([`evmc_message::gas`]).
+      3. If _gas left_ is negative,
+         abort execution with the [`EVMC_OUT_OF_GAS`] status code.
+      4. Otherwise,
+         execute the code of the precompiled contract,
+         return the [`EVMC_SUCCESS`] status code, the output and _gas left_
+         ([`evmc_result::output_data`], [`evmc_result::output_size`], [`evmc_result::gas_left`]).
+   3. Otherwise, emulate execution of empty code by returning
+      the [`EVMC_SUCCESS`] status code
+      and _gas left_ equal the call gas limit ([`evmc_message::gas`]).
+
+Precompiled contract implementations are allowed to return two more EVMC error codes:
+- [`EVMC_FAILURE`] if the failure was caused due to something other than out of gas (e.g. input validation error)
+- [`EVMC_REVERT`] if the precompile doesn't want to forfeit all supplied gas (as of May 2019 no such precompile exists)
+
+The Client is not required to provide the Host interface (by setting the [`evmc_context`] argument of [`execute()`] to null).
+Therefore, the precompiled contracts implementation MUST NOT access the `evmc_context`.
+
+
+## Rationale
+
+It is very unlikely that any precompile will need to access or modify a contract state.
+Not requiring the Client to implement the EVMC Host interface removes the big portion of work
+needed for full EVMC integration.
+
+
+## Test Cases
+
+EVMC provides the [evmc-vmtester] tool for checking compatibility with the EVMC specification.
+
+## Implementations
+
+- [Example of Precompiles VM implementation][example_precompiles_vm.cpp]
+- [ewasm precompiles]
+- Aleth code for precompiles
+- Parity code for precompiles
+
+
+## Copyright
+
+Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
+
+## References
+
+- [EVMC – Ethereum Client-VM Connector API][EVMC]
+- [EVMC documentation]
+- [EVMC VM Implementation Guide][EVMC VM API]
+- [EIP 1352: Specify restricted address range for precompiles/system contracts][EIP-1352]
+
+## TODO
+
+- What if message::kind is CREATE?
```

Comment:
I'd say it should only accept `CALL` with or without the static flag. Everything else is a failure.

Reply:
Ok, I will make it like that except it rather should be a rejection -> the client will have to resolve the problem. Otherwise the failure will be interpreter as if the execution was performed.